### PR TITLE
[BUGFIX beta] Various Chrome deopt fixes.

### PR DESCRIPTION
**Pull request by rwjblue**
- Prevent indexing into primitive types in `guidFor`. Doing `obj[GUID_KEY]` on primitive objects (strings, booleans, numbers, etc) will not work, and triggers a deopt.
- Prevent deopting `Mixin.detect` for each type found. With the truthiness check that existed before, this method is deopted everytime it sees a new type of value. (See screenshots from irhydra in commit message).
- Check `typeof` before testing for truthiness. Using the `typeof` check before the truthiness check, prevents deopts in Chrome for each new object type that is seen (after it initially optimizes).

### CODE REVIEW

File: packages/ember-metal/lib/utils.js [hunk 1]
```diff
@@ -179,7 +179,8 @@ export function generateGuid(obj, prefix) {
   @return {String} the unique guid for this instance.
 */
 export function guidFor(obj) {
-  if (obj && obj[GUID_KEY]) {
+  let type = typeof obj;
+  if ((type === 'object' || type === 'function') && obj !== null && obj[GUID_KEY]) {
```
Comment (krisselden) [inline packages/ember-metal/lib/utils.js]: (type === 'object' && obj !== null) || type === 'function'

**Comment by krisselden** [comment]
@rwjblue you should do the isAngle out of bounds one for the LTS

**Comment by rwjblue** [comment]
Good point @krisselden, submitted https://github.com/emberjs/ember.js/pull/13785 for that.

File: packages/ember-metal/lib/utils.js (hunk 1 above)
Comment (rwjblue) [inline packages/ember-metal/lib/utils.js]: good point, I refactored that a bit to avoid the `!== null` check for functions.

**Comment by rwjblue** [comment]
Confirmed that the `guidFor` and `Mixin.detect` deopts addressed here are fixed (via `initial-render-perf` repo, and irhyrda).  There is still a "wrong map" deopt for `guidFor` (which will continue to happen until we move the GUID into a WeakMap).

**Comment by rwjblue** [comment]
### Before

screenshot

### After

screenshot

**Comment by krisselden** [comment]
@rwjblue that complex list benchmark is pretty bogus, we need to be benchmarking real apps.

**Comment by rwjblue** [comment]
KK, will test with chrome-tracing and report back...

**Comment by krisselden** [comment]
@rwjblue these seem fine, you won't see benefit likely until the downstream deopts are fixed, guidFor needs the define non enumerable property path warmed since it optimizes before hitting it in views.

File: packages/ember-glimmer/lib/utils/references.js [hunk 2]
```diff
@@ -110,7 +110,7 @@ export class PropertyReference extends CachedReference { // jshint ignore:line
       _parentObjectTag.update(tagFor(parentValue));
     }
 
-    if (parentValue && typeof parentValue === 'object') {
+    if (typeof parentValue === 'object' && parentValue) {
```
Comment (stefanpenner) [inline packages/ember-glimmer/lib/utils/references.js]: rather then coercion we can be explicit, the only case we care about here is `parentValue !== null`

**Comment by stefanpenner** [comment]
This looks good to me.

**Comment by stefanpenner** [comment]
@homu r+

**Comment by stefanpenner** [comment]
@homu retry

**Comment by stefanpenner** [comment]
lets assume homu is dead..

**Pull request merged**